File: v1/scala/difference-of-squares/src/main/scala/Squares.scala
class Squares {
  private def square(n: Int) = n * n

  def squareOfSums(n: Int) = square((1 to n).sum)
  def sumOfSquares(n: Int) = (1 to n).map(square).sum
  def difference(n: Int) = squareOfSums(n) - sumOfSquares(n)
}

object Squares {
  def apply() = new Squares()
}

File: v1/haskell/difference-of-squares/Squares.hs
module Squares (sumOfSquares, squareOfSums, difference) where

sumOfSquares :: Integral a => a -> a
sumOfSquares num = sum $ map square [1..num]

squareOfSums :: Integral a => a -> a
squareOfSums num = square $ sum [1..num]

difference :: Integral a => a -> a
difference num = squareOfSums num - sumOfSquares num

square :: Integral a => a -> a
square num = num * num